# Advanced Calculus with Financial Engineering Applications 
# Homework 6
### *Jose Ferreira*
----------------------------

**(7) The yield of a two year quarterly coupon bond with coupon rate 8% is 9%.**

**(i) What are the price B, duration D, and convexity C of the bond?**

In [10]:
import numpy as np

def bond_price_duration_convexity(yield_k, cf_times, cf_values, freq):
    n = np.minimum(len(cf_times), len(cf_values))
    B = 0
    D = 0
    C = 0
    for i in range(0, n):
        B = B + cf_values[i]*(1+(yield_k/freq))**(-freq*cf_times[i])
        D = D + cf_times[i]\
                *cf_values[i]\
                *(1+(yield_k/freq))**(-freq*cf_times[i]-1)
        C = C + cf_times[i]\
                *cf_values[i]\
                *(cf_times[i] + 1/freq)\
                *(1+(yield_k/freq))**(-freq*cf_times[i]-2)
    return B, D/B, C/B

In [4]:
t_cf = [0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2]
v_cf = [2, 2, 2, 2, 2, 2, 2, 102]
freq = 4
yld = 0.09

print ("=================================================")
B, D, C = bond_price_duration_convexity(yld, t_cf, v_cf, freq)
print ("Price = " + str(B))
print ("Modified Duration = " + str(D))
print ("Convexity = " + str(C))

Price = 98.18820384835145
Modified Duration = 1.8254526399515039
Convexity = 3.9232460429683447


**(ii) Use both the formula
$$\frac{B_{new, D}-B}{B}=\frac{\Delta B}{B}\approx -D\Delta y$$
which does not include any convexity adjustment, and the formula 
$$\frac{B_{new, D, C}-B}{B}=\frac{\Delta B}{B}\approx -D\Delta y + \frac{1}{2}C(\Delta y)^2$$
to find an approximate price of the bond if the yield increases by ten basis points (i.e. 0.001), fifty basis points, one percent, two percent and four percent respectively. 
Also, for each different value of the yield change, compute the actual price $B(y + \Delta y)$ of the bond.

We can calculate the new prices with the following equations:
$$B_{new,D} = B(1-D\Delta y)$$
$$B_{new,D, C} = B(1-D\Delta y + \frac{1}{2}C(\Delta y)^2)$$

In [5]:
yield_deltas = np.array([0.001, 0.005, 0.01, 0.02, 0.04])
duration_price_change = -D*yield_deltas
print(duration_price_change)

[-0.00182545 -0.00912726 -0.01825453 -0.03650905 -0.07301811]


In [6]:
B_new_D = B*(1+duration_price_change)
print(B_new_D)

[ 98.00896593  97.29201427  96.39582469  94.60344553  91.01868721]


In [7]:
convexity_price_change = (C/2)*(yield_deltas**2)
print(convexity_price_change)

[  1.96162302e-06   4.90405755e-05   1.96162302e-04   7.84649209e-04
   3.13859683e-03]


In [8]:
B_new_D_C = B*(1 + duration_price_change + convexity_price_change )
print(B_new_D_C)

[ 98.00915854  97.29682947  96.41508551  94.68048883  91.3268604 ]


As for $B(y + \Delta y)$, assuming $\Delta t$ is small enough to ignore its effect: 

In [9]:
actual_price_changes = []
for delta in yield_deltas:
    B, D, C = bond_price_duration_convexity(yld + delta, t_cf, v_cf, freq)
    actual_price_changes.append(B)
print(actual_price_changes)

[98.00915838564238, 97.296810148463862, 96.414931416261894, 94.679264192519696, 91.317191355648262]



| $\Delta y$ 	| $B_{new, D}$ 	| $B_{new, D, C}$ 	| $B(y + \Delta y)$) 	|
|------------	|--------------	|-----------------	|--------------------	|
| 0.001      	|98.00896593   	|98.00915854      	|98.00915838564238   	|
| 0.005      	|97.29201427   	|97.29682947      	|97.296810148463862  	|
| 0.01       	|96.39582469   	|96.41508551      	|96.414931416261894  	|
| 0.02       	|94.60344553   	|94.68048883      	|94.679264192519696  	|
| 0.04       	|91.01868721   	|91.3268604       	|91.317191355648262  	|

**(iii) For each different value of the yield change considered above, compute the relative approximation errors, i.e.,
\begin{align}
\frac{\left|B_{new,D} - B(y + \Delta y)\right|}{B(y + \Delta y)} && \text{and} && \frac{\left|B_{new,D,C} - B(y + \Delta y)\right|}{B(y + \Delta y)}\\
\end{align}
of the approximations above.**

In [16]:
only_duration_error = np.abs(B_new_D - actual_price_changes)/actual_price_changes
print(only_duration_error)

[  1.96362484e-06   4.92912331e-05   1.98171869e-04   8.00794803e-04
   3.26887128e-03]


In [17]:
d_c_error = np.abs(B_new_D_C - actual_price_changes)/actual_price_changes
print(d_c_error)

[  1.58172057e-09   1.98632205e-07   1.59826831e-06   1.29345506e-05
   1.05884130e-04]



| $\Delta y$ 	| $B_{new, D}$ 	| $B_{new, D, C}$ 	| $B(y + \Delta y)$) 	|$\frac{\left|B_{new,D} - B(y + \Delta y)\right|}{B(y + \Delta y)}$ | $\frac{\left|B_{new,D,C} - B(y + \Delta y)\right|}{B(y + \Delta y)}$ |
|------------	|--------------	|-----------------	|--------------------	|-------------------------------|-------------------------------|
| 0.001      	|98.00896593   	|98.00915854      	|98.00915838564238   	|1.96362484e-06   	|1.58172057e-09   	|
| 0.005      	|97.29201427   	|97.29682947      	|97.296810148463862  	|4.92912331e-05   	|1.98632205e-07   	|
| 0.01       	|96.39582469   	|96.41508551      	|96.414931416261894  	|1.98171869e-04   	|1.59826831e-06   	|
| 0.02       	|94.60344553   	|94.68048883      	|94.679264192519696  	|8.00794803e-04   	|1.29345506e-05   	|
| 0.04       	|91.01868721   	|91.3268604       	|91.317191355648262  	|3.26887128e-03   	|1.05884130e-04   	|

----------------------------

**(8) You invest \$20 million in a bond with duration 3 and convexity 18 and \$50 million in a bond with duration 4 and convexity 20.**

**(i) What are the dollar duration and dollar convexity of your portfolio?**

$D_{\$}(V) = B_1 D_1 + B_2 D_2 = \$20.000.000 \times 3 + \$50.000.000 \times 4 = \$260.000.000$   
$C_{\$}(V) = B_1 C_1 + B_2 C_2 = \$20.000.000 \times 18 + \$50.000.000 \times 20 = \$1,360.000.000$ 

**(ii) If the yield curve moves up by 25 bps, find an approximate value of the portfolio**

The change in the value of the portfolio can be approximated as:

$\Delta V \approx -D_{\$}(V)\delta r + \frac{C_{\$}(V)}{2}(\delta r)^2$

For $\delta r = 0.0025$:

$\Delta V \approx -\$260.000.000 \times 0.0025 + \frac{\$1,360.000.000}{2}(0.0025)^2$

$\Delta V = -\$650.000 + \$4.250 = -\$645.750$


**(iii) You can buy or sell two other bonds, one with duration 2 and convexity 7 and another one with duration 4 and convexity 11. What positions could you take in these bonds to immunize your portfolio (i.e. to obtain a portfolio with zero dollar duration and dollar convexity)?**

Let $\Pi$ be the value of the immunized portfolio such as $\Pi = V + B_3 + B_4$, then

$D_{\$}(\Pi) = D_{\$}(V) + B_3 D_3 + B_4 D_4$ 

$D_{\$}(\Pi) = \$260.000.000 + B_3 \times 2 + B_4 \times 4$ 

$C_{\$}(\Pi) = C_{\$}(V) + B_3 C_3 + B_4 C_4$ 

$C_{\$}(\Pi) = \$1,360.000.000 + B_3 \times 7 + B_4 \times 11$ 

For the portfolio to be immunized, $D_{\$}(\Pi) = 0$ and $C_{\$}(\Pi) = 0$. The solution to the system of equations is:
\begin{align}
B_3 = -\$430.000.000; && B_4 = \$150.000.000\\
\end{align}

Therefore, the immunized portfolio should consist of the following positions:
- long \$20 million in $B_1$;
- long \$50 million in $B_2$;
- short \$430 million in $B_3$;
- long \$150 million in $B_4$;
- cash \$280 million.

----------------------------

**(9) Consider a bond portfolio worth \$50mil with DV01 equal to \$10,000 and dollar convexity equal to \$400mil.**

**(i) Assume that the yield curve moves up by fifty basis points. What is the new value of your bond portfolio**